iroh-net/src/discovery.rs

```diff
@@ -95,6 +95,125 @@ pub trait Discovery: std::fmt::Debug + Send + Sync {
+    Unsubscribe(u64),
+    Event(DiscoveryItem),
+}
+
+impl Drop for DiscoveryEvents {
+    fn drop(&mut self) {
+        self.0.handle.abort();
+    }
+}
+
+impl DiscoveryEvents {
+    pub fn new() -> Self {
+        let (sender, mut recv) = channel(20);
+        let event_loop = async move {
+            loop {
+                let mut subscribers: HashMap<u64, Sender<DiscoveryItem>> = HashMap::default();
+                let msg = recv.recv().await;
+                match msg {
+                    Some(Message::Subscribe((id, subscriber))) => {
+                        trace!("Message::Subscribe - Adding new subscriber to discovery events");
+                        subscribers.insert(id, subscriber);
+                    }
+                    Some(Message::Unsubscribe(id)) => {
+                        trace!("Message::Unsubscribe - Removing subscriber {id}");
+                        subscribers.remove(&id);
+                    }
+                    Some(Message::Event(discovery_item)) => {
+                        let mut bad_subscribers = vec![];
+                        for (id, subscriber) in &subscribers {
+                            if let Err(e) = subscriber.send(discovery_item.clone()).await {
+                                error!("Message::Event - Subscriber {id} dropped: {e}");
+                                bad_subscribers.push(id.clone());
+                            }
+                        }
+                        for bad_subscriber in bad_subscribers {
+                            subscribers.remove(&bad_subscriber);
+                        }
+                    }
+                    None => {}
+                }
+            }
+        };
+
+        let handle = tokio::spawn(event_loop);
+        Self(Arc::new(Inner {
+            handle,
+            sender,
+            last_id: AtomicU64::new(0),
+        }))
+    }
+
+    pub async fn send_event(&self, discovery_item: DiscoveryItem) {
+        self.0
+            .sender
+            .send(Message::Event(discovery_item))
+            .await
+            .ok();
+    }
+
+    /// Allows you to subscribe to messages from the `DiscoveryEvent`
+    ///
+    /// Returns a stream of `DiscoveryItems` and an optional unsubscribe
+    /// function.
+    ///
+    /// If a stream is dropped, the `DiscoveryEvent` struct will clean
+    /// up the dangling subscription, even if you do not unsubscribe.
+    pub fn subscribe(
+        &self,
+    ) -> (
+        BoxStream<DiscoveryItem>,
+        impl std::future::Future<Output = ()>,
+    ) {
+        let (send, recv) = channel(20);
+        let id = self
+            .0
+            .last_id
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+        let subscribe_sender = self.0.sender.clone();
+        tokio::spawn(async move {
+            subscribe_sender
+                .send(Message::Subscribe((id, send)))
+                .await
+                .ok();
+        });
```

Comment:
Okay, so you are saying that because we create a task for every time `subscribe` is called, we could be creating an unbounded number of tasks, each of which would be waiting for `subscribe_sender` to have space? So it's better to make `subscribe_sender` unbounded & use `try_send`, so we don't need to `await` the `subscribe_sender.send()` & no longer need to spawn a task?

Just making sure I follow.

Reply:
I'd prefer if this was using `futures_util`'s [`flatten_stream`](https://docs.rs/futures-util/latest/futures_util/future/trait.FutureExt.html#method.flatten_stream) and thus do the `send` as part of the first action in the stream and avoid `tokio::spawn` and `try_send` altogether.

Reply:
This is much prettier looking, at minimum 😂 

To ensure I understand: this basically would defer sending the `Message::Subscribe` event until the stream is `await`d, correct?

Reply:
Yeah.